### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Adjunction): the left partial adjoint

### PR DESCRIPTION
**Pull request by joelriou**
Given a functor `F : D ⥤ C`, we define a functor`F.partialLeftAdjoint : F.PartialLeftAdjointSource ⥤ D` which is defined on a certain full subcategory of `C`. It satisfies similar properties to the left adjoint of `F` (if this existed). We show that the domain of definition of this partial left adjoint is stable under certain colimits.

This shall be used in #17366 in order to show the existence of the pullback functor on categories of presheaves of modules.

---
- [x] depends on: #17389

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Adjunction/PartialAdjoint.lean [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+variable {F}
+
+lemma isRightAdjoint_of_leftAdjointObjIsDefined_eq_top
+    (h : F.LeftAdjointObjIsDefined = ⊤) : F.IsRightAdjoint := by
+  replace h : ∀ X, IsCorepresentable (F ⋙ coyoneda.obj (op X)) := fun X ↦ by
+    simp only [← leftAdjointObjIsDefined_iff, h, Pi.top_apply, Prop.top_eq_true]
+  exact (Adjunction.adjunctionOfEquivLeft
+    (fun X Y ↦ (F ⋙ coyoneda.obj (op X)).corepresentableBy.homEquiv)
+    (fun X Y Y' g f ↦ by apply CorepresentableBy.homEquiv_comp)).isRightAdjoint
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Adjunction/PartialAdjoint.lean:121]: Would be nice to have the converse if it isn't too hard.

**Review by erdOne — APPROVED** [review]
LGTM
Do you have plans to dualize it? 
Or is the right-adjoint version not useful?

File: Mathlib/CategoryTheory/Adjunction/PartialAdjoint.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Adjunction/PartialAdjoint.lean:129]: Thanks for the suggestion, I have added the iff lemma in https://github.com/leanprover-community/mathlib4/pull/17388/commits/7ca7077d174a78fc13926b8aa9799f7b134a97ca

**Comment by joelriou** [comment]
> Do you have plans to dualize it? Or is the right-adjoint version not useful?

I do not know a concrete application of the dual result, but I have added a TODO in the docstring of the file.

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**